### PR TITLE
Enable configuration of monitoring sinks.

### PR DESCRIPTION
**Pull request by vjkoskela**

### CODE REVIEW

File: config/config.conf [hunk 1]
```diff
@@ -13,8 +13,29 @@ httpPort=7090
...
+#    dispatchErrorLoggingInterval = "PT1M"
+#    unsupportedDataLoggingInterval = "PT1M"
+#  }
+#]
 #jvmMetricsCollectionInterval="PT1.0S"
 
+# NOTES:
+# - metricsClientHost and metricsClientPort are deprecated
+# - metricsClientHost and metricsClientPort no longer default from httpHost and httpPort
+# - you CANNOT specify either these AND monitoringSinks
```
Comment (cwbriones) [inline config/config.conf:35]: This wording is a bit awkward, specifically the "either and" construction.

Perhaps something more like "If either metricsClientHost or metricsClientPort are specified, then the legacy configuration will take precedence over monitoringSinks"

File: src/main/java/com/arpnetworking/metrics/mad/configuration/AggregatorConfiguration.java [hunk 2]
```diff
@@ -337,6 +375,14 @@ public Builder setAkkaConfiguration(final Map<String, ?> value) {
         @NotEmpty
         private String _monitoringService = "mad";
         @NotNull
+        private ImmutableList<JsonNode> _monitoringSinks;
+        @NotEmpty
+        private String _metricsClientHost;
```
Comment (cwbriones) [inline src/main/java/com/arpnetworking/metrics/mad/configuration/AggregatorConfiguration.java:380]: Should this and `metricsClientPort` be marked as `@Nullable` now? The `ofNullable` above isn't correct upon inspection without this

File: src/main/java/com/arpnetworking/metrics/mad/Main.java [hunk 3]
```diff
@@ -302,6 +316,36 @@ private Injector launchGuice(final ActorSystem actorSystem) {
...
+        final List<com.arpnetworking.metrics.Sink> sinks = new ArrayList<>();
+        for (final JsonNode sinkNode : monitoringSinks) {
+            @Nullable final JsonNode classNode = sinkNode.get("class");
+            try {
+                if (classNode != null) {
+                    final Class<?> builderClass = Class.forName(classNode.textValue() + "$Builder");
+                    final Object builder = OBJECT_MAPPER.treeToValue(sinkNode, builderClass);
+                    @SuppressWarnings("unchecked")
+                    final com.arpnetworking.metrics.Sink sink =
+                            (com.arpnetworking.metrics.Sink) builderClass.getMethod("build").invoke(builder);
```
Comment (cwbriones) [inline src/main/java/com/arpnetworking/metrics/mad/Main.java:337]: 🤢 

I get why this hack is necessary, it's just really ugly.

File: src/main/java/com/arpnetworking/metrics/mad/parsers/CollectdJsonToRecordParser.java [hunk 4]
```diff
@@ -280,7 +281,7 @@ private CollectdRecord(final Builder builder) {
          * {@link com.arpnetworking.commons.builder.Builder} implementation for
          * {@link CollectdRecord}.
          */
-        public static final class Builder extends ThreadLocalBuilder<CollectdRecord> {
+        public static final class Builder extends OvalBuilder<CollectdRecord> {
```
Comment (cwbriones) [inline src/main/java/com/arpnetworking/metrics/mad/parsers/CollectdJsonToRecordParser.java:284]: unclear why this changed

File: src/main/java/com/arpnetworking/metrics/mad/parsers/GraphitePlaintextToRecordParser.java [hunk 5]
```diff
@@ -219,7 +220,7 @@ private GraphitePlaintextToRecordParser(final Builder builder) {
      * {@link com.arpnetworking.commons.builder.Builder} implementation for
      * {@link GraphitePlaintextToRecordParser}.
      */
-    public static final class Builder extends ThreadLocalBuilder<GraphitePlaintextToRecordParser> {
+    public static final class Builder extends OvalBuilder<GraphitePlaintextToRecordParser> {
```
Comment (cwbriones) [inline src/main/java/com/arpnetworking/metrics/mad/parsers/GraphitePlaintextToRecordParser.java:223]: unclear why this changed

File: src/main/java/com/arpnetworking/metrics/mad/parsers/JsonToRecordParser.java [hunk 6]
```diff
@@ -714,7 +715,7 @@ private static Quantity version2gSampleToQuantity(@Nullable final Version2g.Samp
      * {@link com.arpnetworking.commons.builder.Builder} implementation for
      * {@link JsonToRecordParser}.
      */
-    public static final class Builder extends ThreadLocalBuilder<JsonToRecordParser> {
+    public static final class Builder extends OvalBuilder<JsonToRecordParser> {
```
Comment (cwbriones) [inline src/main/java/com/arpnetworking/metrics/mad/parsers/JsonToRecordParser.java:718]: again, unclear why this changed

File: src/main/java/com/arpnetworking/metrics/mad/Main.java (hunk 3 above)
Comment (vjkoskela) [reply src/main/java/com/arpnetworking/metrics/mad/Main.java:337]: I started with the preferred approach, but there were some nasty surprises in the metrics client. I put it on hold, and hope to return to it in the future to help undo this hack.

File: src/main/java/com/arpnetworking/metrics/mad/parsers/CollectdJsonToRecordParser.java (hunk 4 above)
Comment (vjkoskela) [reply src/main/java/com/arpnetworking/metrics/mad/parsers/CollectdJsonToRecordParser.java:284]: The parser is instantiated in a way that does not use the ThreadLocalBuilder correctly resulting in warnings in the log. It's not related to this PR; just discovered in the testing of this PR.

In general, TLB should be used in areas where objection creation velocity is high. That is not the case with Parsers or Sinks. Therefore, in these cases it is sufficient to use OB.

File: config/config.conf (hunk 1 above)
Comment (vjkoskela) [reply config/config.conf:35]: Done

File: src/main/java/com/arpnetworking/metrics/mad/configuration/AggregatorConfiguration.java (hunk 2 above)
Comment (vjkoskela) [reply src/main/java/com/arpnetworking/metrics/mad/configuration/AggregatorConfiguration.java:380]: Done.

File: src/main/java/com/arpnetworking/metrics/mad/parsers/CollectdJsonToRecordParser.java (hunk 4 above)
Comment (cwbriones) [reply src/main/java/com/arpnetworking/metrics/mad/parsers/CollectdJsonToRecordParser.java:284]: Makes sense, thanks for clarifiying!

Comment (BrandonArp) [reply src/main/java/com/arpnetworking/metrics/mad/parsers/CollectdJsonToRecordParser.java:284]: I think you changed the wrong one to OvalBuilder here.  I don't think the CollectdJsonToRecordParser class has a builder.

**Pull request merged**